### PR TITLE
feat: openapi request bodies

### LINKED ISSUE #935: expose the operator object params to the user in openapi spec
eg: requestBody, operationId, etc

PS: I am working on it.

### PR DESCRIPTION
**Pull request by VishnuSanal**
## Description

This PR fixes #935

## Summary

This PR does add request bodies to openapi implementation.

## PR Checklist

Please ensure that:

- [x] The PR contains a descriptive title
- [x] The PR contains a descriptive summary of the changes
- [x] You build and test your changes before submitting a PR.
- [x] You have added relevant documentation
- [x] You have added relevant tests. We prefer integration tests wherever possible

## Pre-Commit Instructions:

- [x] Ensure that you have run the [pre-commit hooks](https://github.com/sansyrox/robyn#%EF%B8%8F-to-develop-locally) on your PR.

### CODE REVIEW

File: docs_src/src/pages/documentation/api_reference/openapi.mdx
Comment (sansyrox) [inline docs_src/src/pages/documentation/api_reference/openapi.mdx:247]: The user should not be expected to type in openapi json when typing in the request body. it should either be a class or a typed dictionary

**Review by sansyrox — CHANGES_REQUESTED** [review]
Not super happy with the interface here. @VishnuSanal , we need to allow either a class or a typeddict like in query params

File: docs_src/src/pages/documentation/api_reference/openapi.mdx
Comment (VishnuSanal) [reply docs_src/src/pages/documentation/api_reference/openapi.mdx:247]: hi, isn't this what we meant when we decided we "provide" the necessary params  at first & we "support" all params, if the user wants them?

Comment (sansyrox) [reply docs_src/src/pages/documentation/api_reference/openapi.mdx:247]: @VishnuSanal , would you say the `body` of a request is not a necessary param?

File: robyn/openapi.py
Comment (VishnuSanal) [inline robyn/openapi.py:144]: the openapi `schema` supports all the properties from the [JSON Schema Specification Draft](https://tools.ietf.org/html/draft-bhutton-json-schema-00). ref: [https://swagger.io/specification/#schema-object](https://swagger.io/specification/#schema-object). I don't see a point in us creating a class for this; for we would restrict the user from using all the params. what do you think?

Comment (sansyrox) [reply robyn/openapi.py:144]: @VishnuSanal , we only care about the user's developer experience. It is foolish of us to expect the user to write openapi style json for documentation.

If we want the user to write openpai JSON for the body, why even bother implementing this feature in the first place?

I hope this answers your question.

Comment (VishnuSanal) [reply robyn/openapi.py:144]: > we only care about the user's developer experience. It is foolish of us to expect the user to write openapi style json for documentation.

I can agree; but, my concern was whether we'll restrict the user from using any of the params mentioned in the entirety of the JSON schema draft that the OAS schema key supports.

Rephrasing myself: say, we implement a data class for schema with the most used keys. What if a user wants to use a key that we haven't included? Or, if we choose to add support for **all keys** in the entire schema draft, will it be worth the effort we put in to support all of those keys inside a data class -- for we don't know whether they'll be used or not.

I was opting for flexibility for the user to choose which keys that they want to use by letting the user input the schema as a dict (like we did for all the keys inside the components object).

Please correct me if my assumptions are wrong, or if there's a better way around to get the best of both worlds. Thanks!

> If we want the user to write openpai JSON for the body, why even bother implementing this feature in the first place?
>
> I hope this answers your question.

I am not sure why you are being furious about this.

Comment (sansyrox) [reply robyn/openapi.py:144]: > I am not sure why you are being furious about this.

There are two reasons. I'll explain in tomorrow's meeting

File: docs_src/src/pages/documentation/api_reference/openapi.mdx
Comment (sansyrox) [inline docs_src/src/pages/documentation/api_reference/openapi.mdx:265]: @VishnuSanal , the developer is not supposed to write all this. just a class with key value pairs

File: docs_src/src/pages/documentation/api_reference/openapi.mdx
Comment (sansyrox) [inline docs_src/src/pages/documentation/api_reference/openapi.mdx:277]: this should never be a part of the callback decorator

**Review by sansyrox — CHANGES_REQUESTED** [review]
Hey @VishnuSanal ,

needs fixing.

**Comment by sansyrox** [comment]
The body should resemble the style of query params

**Comment by sansyrox** [comment]
@VishnuSanal , this should be the interface

```
class CustomBody(TypedDict):
    file: bytes
    file_name: str

@app.post("/sync/multipart-file")
def sync_multipart_file(request: Request, body: CustomBody):
    files = request.files
    file_names = files.keys()
    return {"file_names": list(file_names)}
```

File: docs_src/src/pages/documentation/api_reference/openapi.mdx
Comment (sansyrox) [inline docs_src/src/pages/documentation/api_reference/openapi.mdx:234]: @VishnuSanal , this was an example , we should improve the docs with a working example once the code is ready

**Comment by VishnuSanal** [comment]
https://github.com/sparckles/Robyn/actions/runs/10579769894/job/29313024961 please try rerunning. Issue is with websocket tests.

**Comment by sansyrox** [comment]
@VishnuSanal , is this PR up for review?

File: robyn/openapi.py
Comment (sansyrox) [inline robyn/openapi.py:227]: what does this field represent?

**Comment by sansyrox** [comment]
@VishnuSanal , could you please show the example of the openapi page generated?

Attach a screenshot?

**Comment by VishnuSanal** [comment]
> @VishnuSanal , is this PR up for review?

Yep

File: robyn/openapi.py
Comment (VishnuSanal) [reply robyn/openapi.py:227]: We are appending it below. This line is for init. Else, it complained "dict doesn't contain key parameter", IIRC. I'll check.

**Comment by VishnuSanal** [comment]
image

```
class CreateItemBody(TypedDict):
    name: Optional[str]
    description: str
    price: float
    tax: float


@app.post("/upload/file")
def create_item(request: Request, body=CreateItemBody):
    return request.body
```

**Comment by sansyrox** [comment]
Amazing! ✨ 
Next update should be to make them look like the following

```python
class CreateItemBody(TypedDict):
    name: Optional[str]
    description: str
    price: float
    tax: float


@app.post("/upload/file")
def create_item(request: Request, body: CreateItemBody):
    return request.body
```

But we can address that in the next PR when we add support for response schemas! GG @VishnuSanal . Need to give a final review

File: integration_tests/test_openapi.py
Comment (sansyrox) [inline integration_tests/test_openapi.py:60]:
```suggestion
def test_openapi_request_body():
```

File: robyn/openapi.py
Comment (sansyrox) [inline robyn/openapi.py:336]: @VishnuSanal , what happens if it is a custom type?

Comment (VishnuSanal) [reply robyn/openapi.py:336]: > @VishnuSanal , what happens if it is a custom type?

uh, oh -- it defaults to string at present. any other suggestions? meanwhile I will check what fastapi does & update here.

these are the supported data types: https://swagger.io/specification/#data-types

Comment (sansyrox) [reply robyn/openapi.py:336]: I am not sure. Maybe just use the class name if nothing works. Or recursively parse them?

Comment (VishnuSanal) [reply robyn/openapi.py:336]: we get the following if we use an "objject" type.

```{
  "name": {},
  "description": "string",
  "price": 0,
  "tax": 0
}```

@sansyrox I think we can default it to it ("object") instead of a string. what say?

Comment (sansyrox) [reply robyn/openapi.py:336]: @VishnuSanal , should recursively parse it like fastpai

Comment (sansyrox) [reply robyn/openapi.py:336]: @VishnuSanal , did you address this?

Comment (VishnuSanal) [reply robyn/openapi.py:336]: yeps yeps

**Comment by sansyrox** [comment]
Love this @VishnuSanal ✨ 

Doing a final review before merging

**Review by sansyrox — APPROVED** [review]
Looks good to me 😄 

Good job!

**Pull request merged**
